canopen/objectdictionary/datatypes.py

```diff
@@ -10,16 +11,92 @@
 VISIBLE_STRING = 0x9
 OCTET_STRING = 0xA
 UNICODE_STRING = 0xB
+TIME_OF_DAY = 0xC
+TIME_DIFFERENCE = 0xD
 DOMAIN = 0xF
 INTEGER24 = 0x10
 REAL64 = 0x11
+INTEGER40 = 0x12
+INTEGER48 = 0x13
+INTEGER56 = 0x14
 INTEGER64 = 0x15
 UNSIGNED24 = 0x16
+UNSIGNED40 = 0x18
+UNSIGNED48 = 0x19
+UNSIGNED56 = 0x1A
 UNSIGNED64 = 0x1B
+PDO_COMMUNICATION_PARAMETER = 0x20
+PDO_MAPPING = 0x21
+SDO_PARAMETER = 0x22
+IDENTITY = 0x23
 
-SIGNED_TYPES = (INTEGER8, INTEGER16, INTEGER24, INTEGER32, INTEGER64)
-UNSIGNED_TYPES = (UNSIGNED8, UNSIGNED16, UNSIGNED24, UNSIGNED32, UNSIGNED64)
+SIGNED_TYPES = (INTEGER8, INTEGER16, INTEGER24, INTEGER32, INTEGER40, INTEGER48, INTEGER56, INTEGER64)
+UNSIGNED_TYPES = (UNSIGNED8, UNSIGNED16, UNSIGNED24, UNSIGNED32, UNSIGNED40, UNSIGNED48, UNSIGNED56, UNSIGNED64)
 INTEGER_TYPES = SIGNED_TYPES + UNSIGNED_TYPES
 FLOAT_TYPES = (REAL32, REAL64)
 NUMBER_TYPES = INTEGER_TYPES + FLOAT_TYPES
 DATA_TYPES = (VISIBLE_STRING, OCTET_STRING, UNICODE_STRING, DOMAIN)
+
+
+class UnsignedN:
```

Comment:
I just noticed that using these classes (just as the previous 24-bit only variants) would break with PDO that are not aligned to full bytes.  That's because the code there actually examines the `.format` attribute, which is only available on the real `Struct` class:

https://github.com/christiansandberg/canopen/blob/94f337d779c9294db99083821a6a834fe58b167a/canopen/pdo/base.py#L567

Let's add a read-only `str` property to forward `.format` to the `self.struct.format` instance variable.

Reply:
What _exactly_ is this test doing? It tests that the lower-case of the struct format string is not empty? So is it then a method to determine all types that are struct native, i.e. not 24, 48, 56 and so on? This code is very strange to me. The comment sais "Check if the variable is signed and if the data is negative prepend signedness". I think I need some enlightenment here.

Reply:
The test hinges on the implementation detail that the struct mini-language uses uppercase codes (B, H, L, Q) for unsigned and lowercase (b, h, l, q) for signed types.  It then goes on to check whether the value is in the upper half of the representable number space (with the specified number of bits), which indicates that it was a negative number being truncated.  In that case, padding with 1-bits is added to the left to make it a truly negative integer (in two's complement) after converting back to the full-width integer type.

This is needed for partially mapped signed integer variables.  Judging from the name, the `STRUCT_TYPES` listing was meant to indicate we do have a `struct`-based en- / decoder available, thus it can be assumed that the uppercase / lowercase distinction works.

I think the easiest way to fix this is to actually inherit from `struct.Struct` in your generic integer classes, so that we do have access to a real `.format` attribute again.  Inheritance instead of composition, as an OOP specialist would put it.

Reply:
Ahhh. It sais `islower()` not `lower()`... Sorry, I misread it.

I agree, let's implement some upper case letter for unsigned and lower case for signed. I think simply proxying `format` to `self.struct.format` in the `UnsignedN` and `SignedN` object will suffice.

Reply:
Hm. I cannot say I share that experience.  Tests seem to pass alright with the above implementation.

Reply:
If the code is safe, I'd say this is good. -- There are probably other ways to declare `class UnsignedN` to be `struct.Struct`-like so that a `dict[int, Struct]` annotation would work.

Allow me to be direct, @erlend-aasland: Are there any reasons to shy away from inheriting cpython classes, such as `struct.Struct` using the above code?

Reply:
`struct.Struct` is a base type; you can safely inherit from it. Builtin types and stdlib types that are not safe to inherit from are not base types (or _should not_ be base types). See also:

- https://docs.python.org/3.14/c-api/typeobj.html#c.Py_TPFLAGS_BASETYPE

I see now that that page needs some care (yes, the triple question marks...)

Reply:
FTR, the `struct.Struct` type spec can be found here:

https://github.com/python/cpython/blob/4055577221f5f52af329e87f31d81bb8fb02c504/Modules/_struct.c#L2239-L2246

Reply:
Wow, it never crossed my mind that a Python class in the stdlib could NOT be subclassed.  If such a thing exists, I hope there is a big fat runtime or compiler warning, besides being clearly stated in the documentation.

Reply:
You'll get a `TypeError`:

```python
Python 3.13.0b1 (v3.13.0b1:2268289a47, May  8 2024, 06:41:53) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import datetime
>>> class C(datetime.timezone): ...
... 
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    class C(datetime.timezone): ...
TypeError: type 'datetime.timezone' is not an acceptable base type
>>> 
```

For most user-visible types, there is a notice of some kind in the docs:

- https://docs.python.org/3/library/datetime.html#datetime.tzinfo

Reply:
Anyway (continuing the digression): most non-base types are not user visible. For example, the `_struct.unpack_iter` type:

```python
Python 3.13.0b1 (v3.13.0b1:2268289a47, May  8 2024, 06:41:53) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import struct
>>> tp = type(struct.iter_unpack("b", b'x'))
>>> tp.__name__
'unpack_iterator'
>>> class C(tp): ...
... 
Traceback (most recent call last):
  File "<python-input-3>", line 1, in <module>
    class C(tp): ...
TypeError: type '_struct.unpack_iterator' is not an acceptable base type
>>> 
```